include/zephyr/sys/hash_map.h

```diff
@@ -102,6 +104,16 @@ extern "C" {
  */
 #define SYS_HASHMAP_DEFINE_STATIC(_name) SYS_HASHMAP_DEFAULT_DEFINE_STATIC(_name)
 
+/*
+ * Default equality function used by the hash map implementations
+ */
+static inline bool sys_hashmap_default_equal(uint64_t left, uint64_t right) {
```

Comment:
Something is inconsistent between the associated issue and this PR.

Namely, the issue describes using strings as _keys_ for hash maps, but it would appear that this PR aims to use strings as _values_. I might be mistaken though.

Using strings as _values_ should be considerably easier to implement, as the pointer of course will fit inside the 64-bit value for both 32-bit and 64-bit architectures, and I believe that is supported today.

However, the value side of the hash map does not need to be unique. Only the key of the hash map needs to be unique.

If the aim of this PR is to use strings as keys, then I would say a significant problem is that hash maps currently only use 32-bit keys. I.e. on 64-bit architectures, the 64-bit `char *` would not be a valid key. So there would be a need to define 64-bit hash functions, maps, etc.

However, again, with the concept of using strings as keys simply requires not using duplicate keys. So perhaps it would be easier to de-duplicate strings first?

What is a simple way to de-duplicate strings? Well, calculating a 32-bit hash of the string and seeing if the key exists in some collection is one way. Another way is to use a trie. I think I would lean toward the former if the strings were expected to be mostly dissimilar. 

This is typically what would be done in languages that implicitly ref-count strings such as C++ (and others), have smart pointers, automatic destruction, etc.

What might be convenient to do in this case is to define a separate specialization of the hash map just for C strings, since there is that additional step of deduplicating keys (and actually, one way to define a set is to use a hash map).

Maybe a `string_set` and some method like add, remove, contains, and hash or something like that?

I would guess that a key space of 4B is large enough to handle all strings used in Zephyr for most practical applications.

Reply:
What I'm trying to do is make it feasible to store generic data types in the key field of hashmaps of course through the use of pointers. This should still be achieved by the user who is required to provide a custom equality function and a custom hash function that correctly hashes the associated pointer.
I don't get what you're saying about the 32-bit hash not being enough, it's true that currently, the hash function returns a 32-bit number but it seems to me that the key is still stored as an [`uint64_t`](https://github.com/zephyrproject-rtos/zephyr/blob/5305aa615b0f800508e253c3398712eca1b209b4/lib/hash/hash_map_sc.c#L20) so it seems completely possible to store a pointer even if the hash function returns a smaller hash.
I see you also mentioned handling the allocations, and yes that makes things more complicated to handle on the hashmap code part. Currently however users are already storing pointers in the value field of maps and when they do they also handle deallocating them.
A valid use case I see is something like the following:
```c
#include <zephyr/sys/hash_function.h>
#include <zephyr/sys/hash_map.h>
#include <zephyr/sys/hash_map_api.h>
#include <zephyr/sys/hash_map_sc.h>

// NOLINTNEXTLINE(hicpp-signed-bitwise)
const struct sys_hashmap_config object_map_config
    = SYS_HASHMAP_CONFIG(OBJECT_MAP_MAX_ENTRIES, SYS_HASHMAP_DEFAULT_LOAD_FACTOR);

// we check that we can cast correctly to uint64_t
BUILD_ASSERT(sizeof(uintptr_t) <= sizeof(uint64_t), "A pointer is larger than 64 bits ?!");

// NOLINTNEXTLINE(hicpp-signed-bitwise)
const struct sys_hashmap_config object_map_config
    = SYS_HASHMAP_CONFIG(OBJECT_MAP_MAX_ENTRIES, SYS_HASHMAP_DEFAULT_LOAD_FACTOR);

void example(...) {
    struct sys_hashmap_data entries_map_data = {};
    struct sys_hashmap entries_map = {
        .api = &sys_hashmap_sc_api,
        .config = &object_map_config,
        .data = &entries_map_data,
        // only pointer to valid c strings should be hashed
        .hash_func = hash_string,
        .eq_func = eq_string,
        .alloc_func = SYS_HASHMAP_DEFAULT_ALLOCATOR,
    };

    // use the map

exit:
    sys_hashmap_clear(&entries_map, NULL, NULL);
    return result;
}

static uint32_t hash_string(const void *uint64_ptr_key, size_t len) {
    CHECK_HALT(len != 8, "Received an invalid length for the hasmap key, expected only uint64_t");

    uint64_t key_pointer = {};
    memcpy(&key_pointer, uint64_ptr_key, sizeof(uint64_t));

    // we know (and hope) this is actually a string
    // NOLINTNEXTLINE(performance-no-int-to-ptr)
    const char *key_string = UINT_TO_POINTER(key_pointer);
    size_t key_string_len = strlen(key_string);

    //LOG_INF("Hashing '%s' = '%016x'", key_string, sys_hash32(key_string, key_string_len));

    return sys_hash32(key_string, key_string_len);
}

static bool eq_string(uint64_t key_left, uint64_t key_right) {
    const char *key_left_string = UINT_TO_POINTER(key_left);
    const char *key_right_string = UINT_TO_POINTER(key_right);

    return strcmp(key_left_string, key_right_string) == 0;
}
```
The hashmap does not handle anything besides what it always did but by exposing the `eq_func` field the user can override the default comparison logic and compare the keys.
What I got about not using duplicated strings is that if all the pointers used to store the string `"example"` pointed to the same instance (same pointer) we wouldn't need any fancy equality function but this is more complicated in other ways.
Currently, I do not see why the simple equality function override method would not work, but I'm new to the field and I'm probably missing something.
For the same string, you get the same hash (given that the hash function is defined correctly) after that the correct bucket will be checked (in the case of sc) every entry key stored in the bucket will be compared to the key passed (another pointer that could point to the same string) using the equality function.

Reply:
> What I'm trying to do is make it feasible to store generic data types in the key field of hashmaps of course through the use of pointers. This should still be achieved by the user who is required to provide a custom equality function and a custom hash function that correctly hashes the associated pointer.
> I don't get what you're saying about the 32-bit hash not being enough, it's true that currently, the hash function returns a 32-bit number but it seems to me that the key is still stored as an [`uint64_t`](https://github.com/zephyrproject-rtos/zephyr/blob/5305aa615b0f800508e253c3398712eca1b209b4/lib/hash/hash_map_sc.c#L20) so it seems completely possible to store a pointer even if the hash function returns a smaller hash.

That's correct - sorry, I was looking at this originally.

https://github.com/zephyrproject-rtos/zephyr/blob/main/include%2Fzephyr%2Fsys%2Fhash_map.h#L132-L133

It looks like the key type was left as uint64_t intentionally. It's been a while.

> I see you also mentioned handling the allocations, and yes that makes things more complicated to handle on the hashmap code part. Currently however users are already storing pointers in the value field of maps and when they do they also handle deallocating them.

As long as key allocations are handled externally, that should be ok. Currently only value allocations are managed.

> The hashmap does not handle anything besides what it always did but by exposing the `eq_func` field the user can override the default comparison logic and compare the keys.

Sure - I think it should be doable. Can you make the necessary changes in this PR?

There would need to be new tests added.

Reply:
Ok, thanks again for your time. I will try to clean the pr up and add a few tests, maybe even some additions to the documentation would be a good idea.